Flexvolume resize implementation

aniket-s-kulkarni opened the pull request:
**What this PR does / why we need it**

This PR adds support for expanding Flex Persistent Volumes - kubernetes/features#304

- This PR allows the expansion code flow to execute on the kubelet for flex volumes rather than the controller as the flex plugins are not installed on the controller.
- It allows the flex volume plugin to implement ExpandVolumeDevice method as well as a new ExpandFS method to allow for expanding in use Flex PersistentVolumes
- Includes e2e tests

```release-note
Flex volume plugins now support expandvolume (to increase underlying volume capacity) and expanfs (resize filesystem) commands that Flex plugin authors can implement to support expanding in use Flex PersistentVolumes
```

/sig storage
/assign @gnufied @chakri-nelluri

Commit:
adding support for expanding in use persistent volumes for Flex

Commit:
e2e tests for flex volume expand

mysterious build error

neolit123 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1265,26 +1281,54 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
+					pvcWithResizeRequest.CurrentSize)
+
+				if expandErr != nil {
+					detailedErr := fmt.Errorf("error expanding volume %q of plugin %s: %v", pvcWithResizeRequest.QualifiedName(), volumePlugin.GetPluginName(), expandErr)
```
`...of plugin %q...`

neolit123 commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,21 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
+		if volumeSpec := volume.NewSpecFromPersistentVolume(pv, false); volumeSpec.PersistentVolume.Spec.FlexVolume == nil {
+			// A flex volume is expandable but since it resides on the kubelet not the controller
+			// and is dynamically probed plugin, FindExpandablePluginBySpec will never find it
+			// Therefore explicitly allow flex volumes to resize
```
please add punctuation in the comment.

`...find it.`
`Therefore explicitly allow flex volumes to resize.`

the comment seems to repeat multiple times in the diff.

neolit123 commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,21 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
+			// and is dynamically probed plugin, FindExpandablePluginBySpec will never find it
+			// Therefore explicitly allow flex volumes to resize
+			volumePlugin, err := expc.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
+			glog.Warningf("plugin=%v err=%v fv=%v", volumePlugin != nil, err, volumeSpec.PersistentVolume.Spec.FlexVolume)
```
should this be without `glog.V(x)`?

neolit123 commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -287,3 +291,34 @@ func (plugin *flexVolumePlugin) getDeviceMountPath(spec *volume.Spec) (string, e
+	call.Append(deviceMountPath)
+
+	_, err := call.Run()
+
```
please remove this empty line.

neolit123 commented on pkg/volume/plugins.go:
```diff
@@ -841,6 +848,34 @@ func (pm *VolumePluginMgr) FindMapperPluginByName(name string) (BlockVolumePlugi
+	if err != nil {
+		return nil, err
+	}
+
```
best to trim some of the empty lines that follow a line that only has `}`

neolit123 reviewed (commented):
@aniket-s-kulkarni 
thank you for your PR!
added a couple of small comments about formatting.

aniket-s-kulkarni commented:
@neolit123 I made the suggested changes.

brahmaroutu commented:
Docs: https://github.com/kubernetes/website/pull/10097
@zparnold

dims commented:
/ok-to-test

neolit123 commented:
/uncc

aniket-s-kulkarni commented:
/ok-to-test

aniket-s-kulkarni commented:
/retest

linyouchong commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,21 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
-				util.GetPersistentVolumeClaimQualifiedName(newPVC), newPVC.UID, err)
-			return
+		if volumeSpec := volume.NewSpecFromPersistentVolume(pv, false); volumeSpec.PersistentVolume.Spec.FlexVolume == nil {
+			// A flex volume is expandable but since it resides on the kubelet not the controller
```
the volume driver shall be installed on master node.
According to https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/grow-flexvolume-size.md#expandfs
```
If flex volume driver supports resizing, it shall implement expandvolume method and at least, the volume driver shall be installed on master node.
```

gnufied replied:
Did we finalize on that? I don't particularly like special casing here for flex. Can we not make it a hard requirement that for flex volume resizing the plugin must be at least installed on the controller so as controller is aware of plugin capabilities before allowing volume expansion?

aniket-s-kulkarni replied:
@gnufied I was under the impression there was discussion around not make that (installing flex plugin on master) a hard requirement. Maybe we need to revisit it? 

If we do make it a hard requirement, it would lead to issues for certain uses cases. Esp. for "managed Kubernetes service" like [Amazon EKS](https://aws.amazon.com/eks/) or [IBM CKS](https://www.ibm.com/cloud/container-service) or [AKS](https://azure.microsoft.com/en-us/services/kubernetes-service) - in each cases the control management is hidden from the end user making it tough to install the flex driver on the master.

How about an alternative -

1. The volume-expand controller sets the PVC condition to `ExternalResizing` instead of `Resizing` for plugins (only Flex today) that don't reside on the master/controller
2. The kubelet reconciler calls `ExpandVolumeDevice` of the plugin (Flex) and then 
  - if it succeeds the condition is cleared prompting the volume-expand controller to finish and update PV and move onto fs resizing as it does today
  - if it fails the condition is updated to `ExternalResizingFailed` which prompts the volume-expand to abort resizing with appropriate error

This kind of controller <-> kubelet feedback loop is necessary since only the controller is guaranteed to have privileges to update PV, the kubelet may not be able to do it.

gnufied replied:
@saad-ali @chakri-nelluri what you guys think about this? I think that - controller to kubelet and then back to controller feedback loop is problematic. Expand Controller must reprocess such PVCs.

aniket-s-kulkarni replied:
@saad-ali @chakri-nelluri - @gnufied and I were discussing this on slack, and there maybe a simpler way.

This discussion got underway because the design mentioned the need to install Flex on the controller so that a Flex author can signal expansion is unsupported by returning an error from `ExpandVolumeDevice`. 

Since only dynamically provisioned volumes with `allowVolumeExpansion = true` in their storage classes are allowed to be expanded, a Flex plugin that doesn't support expansion can simply set `allowVolumeExpansion = false` and we don't need to come up with a complicated flow to handle it.

So is that a good enough option?

chakri-nelluri replied:
@aniket-s-kulkarni if the plugin is not installed on master, assume it to be true and mark the volume "RequiresFSResize" and let the Kubelet side code manage it. Let me know  if you see any issues with it?

chakri-nelluri replied:
@aniket-s-kulkarni The code freeze is very close. Ping me on slack if you need a quick review.

aniket-s-kulkarni replied:
@chakri-nelluri you are right. That works - and that's how the PR is doing it. And with what @gnufied mentioned above - it covers all the cases. I will address your comment about abstracting the `.FlexVolume != nil` checks

linyouchong commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,21 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
+			// Therefore explicitly allow flex volumes to resize
+			volumePlugin, err := expc.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
+			glog.V(4).Infof("plugin=%v err=%v fv=%v", volumePlugin != nil, err, volumeSpec.PersistentVolume.Spec.FlexVolume)
+			if err != nil || (volumePlugin == nil && volumeSpec.PersistentVolume.Spec.FlexVolume == nil) {
```
`volumeSpec.PersistentVolume.Spec.FlexVolume == nil`   always be `true` at this line, should deleted it.

aniket-s-kulkarni replied:
sounds good.

aniket-s-kulkarni commented:
/retest

chakri-nelluri commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1265,26 +1281,54 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
 	volumePlugin, err := og.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
 
-	if err != nil {
+	if err != nil && volumeSpec.PersistentVolume.Spec.FlexVolume == nil {
```
Is there a way we can abstract Flexvolume type here. We should not bring knowledge of Flexvolume into operationexcutor

chakri-nelluri commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -287,3 +291,33 @@ func (plugin *flexVolumePlugin) getDeviceMountPath(spec *volume.Spec) (string, e
+	call := plugin.NewDriverCall(expandFSCmd)
+	call.AppendSpec(spec, plugin.host, nil)
+	call.Append(devicePath)
+	call.Append(deviceMountPath)
```
We need the new size of the volume.

aniket-s-kulkarni replied:
@chakri-nelluri do we need the new size? The size has already been passed in the prior call (`ExpandVolumeDevice`) and has been acted on by the plugin.

chakri-nelluri replied:
@aniket-s-kulkarni Yes, there can be plugins which can support FS resize, but have no corresponding block resize calls. They expand dynamically and can ignore previous device resize call.

aniket-s-kulkarni replied:
Ok I can look up the size from `spec.PersistentVolume.Status.Capacity` and add that in here for the call.

chakri-nelluri commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -287,3 +291,33 @@ func (plugin *flexVolumePlugin) getDeviceMountPath(spec *volume.Spec) (string, e
+	call.AppendSpec(spec, plugin.host, nil)
+	call.Append(strconv.FormatInt(newSize.Value(), 10))
+	call.Append(strconv.FormatInt(oldSize.Value(), 10))
+
```
Let us also add device path for this call.

aniket-s-kulkarni replied:
Done

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -632,13 +633,28 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, devi
+			if fsResizablePlugin, callExpandFs := expandableVolumePlugin.(volume.FSResizableVolumePlugin); callExpandFs {
+				// special case for flex volumes
+				if _, resizeErr = fsResizablePlugin.ExpandVolumeDevice(volumeToMount.VolumeSpec, pvSpecCap, pvcStatusCap); resizeErr == nil {
+					resizeErr = fsResizablePlugin.ExpandFS(volumeToMount.VolumeSpec, devicePath, deviceMountPath)
```
`ExpandFS` call should be extended to all volume plugins that require file system resize rather than special casing this for Flex.

chakri-nelluri replied:
@aniket-s-kulkarni Yes this would be required for all plugins going forward.

aniket-s-kulkarni replied:
OK so we go ahead and add this to all existing plugins.

gnufied replied:
yeah. Also all the existing plugins will simply call out to existing util resize function.

chakri-nelluri commented:
@aniket-s-kulkarni The code freeze is very close. Ping me on slack if you need a quick review.

aniket-s-kulkarni commented:
@chakri-nelluri I have addressed your comments

aniket-s-kulkarni commented:
/retest

aniket-s-kulkarni commented:
/retest

aniket-s-kulkarni commented:
/test pull-kubernetes-verify

aniket-s-kulkarni commented:
/test pull-kubernetes-e2e-gce-100-performance

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -393,7 +393,7 @@ func (spec *Spec) Name() string {
 // IsControllerExpandable returns true for volume types that can be expanded by the controller
 // and false for those that need to be run on the node. Flex volume is the only one in the
-// latter category since its typically not installed on the controller/master
+// latter category since its typically not installed on the controller
```
later since it is typically not installed on the controller

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -384,6 +391,36 @@ func (spec *Spec) Name() string {
+	case spec.PersistentVolume != nil:
+		return spec.PersistentVolume.Spec.FlexVolume == nil
+	default:
+		return true
```
return false for default case?

aniket-s-kulkarni replied:
Fair enough - although in such a case (i.e. `.Volume == nil && .PersistentVolume == nil`) something else altogether would be wrong.

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -384,6 +391,36 @@ func (spec *Spec) Name() string {
+// this is used in context on the controller where the plugin lookup fails
+// as volume expansion on controller isn't supported, but a plugin name is
+// required
+func (spec *Spec) NonControllerExpandablePluginName() string {
```
Rename it to "Host or KubeletExpandable" ?

aniket-s-kulkarni replied:
Will change both to `IsKubeletExpandable` and `KubeletExpandablePluginName`

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -384,6 +391,36 @@ func (spec *Spec) Name() string {
+	case spec.Volume != nil:
+		return "flexvolume-" + spec.Volume.FlexVolume.Driver
+	case spec.PersistentVolume != nil:
+		return "flexvolume-" + spec.PersistentVolume.Spec.FlexVolume.Driver
```
Also adding flexvolume as a prefix might blow some naming limits imposed on the objects.

aniket-s-kulkarni replied:
I will just forego the prefix.

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -384,6 +391,36 @@ func (spec *Spec) Name() string {
+// required
+func (spec *Spec) NonControllerExpandablePluginName() string {
+	switch {
+	case spec.Volume != nil:
```
Add check for flexvolume, this would return true for all plugins?

aniket-s-kulkarni replied:
Yes, missed that. Done.

chakri-nelluri commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -632,13 +633,28 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, devi
+			var resizeStatus bool
+			var resizeErr error
+
+			if fsResizablePlugin, callExpandFs := expandableVolumePlugin.(volume.FSResizableVolumePlugin); callExpandFs {
```
*Nit - callExpandFS

chakri-nelluri commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1265,26 +1281,53 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
+		// A flex volume is expandable but since it resides on the kubelet not the controller
+		// and is dynamically probed plugin, FindExpandablePluginBySpec will never find it
+		// Therefore fake the volume expansion on the controller.
+		glog.V(4).Infof("Faking expansion for flex volume %s", pvcWithResizeRequest.QualifiedName())
```
Change this to - No plugin found, so assuming resizable? This can be generic.

aniket-s-kulkarni replied:
OK

chakri-nelluri commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1265,26 +1281,53 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
+					return detailedErr, detailedErr
+				}
+			} else {
+				// this is the case for flex volume, which doesn't reside on the controller
```
Same here - Change the comment to "if plugin is not found, assume it is resizable"

aniket-s-kulkarni replied:
OK

chakri-nelluri commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,21 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
+			volumePlugin, err := expc.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
+			glog.V(4).Infof("plugin=%v err=%v fv=%v", volumePlugin != nil, err, volumeSpec.PersistentVolume.Spec.FlexVolume)
+			if err != nil || volumePlugin == nil {
+				err = fmt.Errorf("didn't find a plugin capable of expanding the volume; " +
```
if volumePlugin==nil, this would not be a error message. Please fix the log level based on err != nil & volumeplugin == nil.

aniket-s-kulkarni replied:
@chakri-nelluri this is existing behavior - if the plugin **is** expandable by the controller, but the plugin object isn't found or there was an error looking up the expandable plugin then in _both_ cases we need this error right? `FindExpandablePluginBySpec` can either return 

- `err == nil && volumePlugin == nil` (i.e. plugin not found)
- Or `err != nil && volumePlugin == nil` (i.e. there was an error in the plugin lookup itself)

So the error needs to be reported `err != nil || volumePlugin == nil` right?

chakri-nelluri replied:
Got it. We will not visit this path in case of flex, since we won't even find the plugin in the earlier logic.

chakri-nelluri commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,21 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
+			if err != nil || volumePlugin == nil {
+				err = fmt.Errorf("didn't find a plugin capable of expanding the volume; " +
+					"waiting for an external controller to process this PVC")
+				expc.recorder.Event(newPVC, v1.EventTypeNormal, events.ExternalExpanding,
```
Same use a different - EventType when err != nil

aniket-s-kulkarni replied:
Sure, will use `v1.EventTypeWarning`

chakri-nelluri commented on pkg/controller/volume/expand/pvc_populator.go:
```diff
@@ -99,16 +99,20 @@ func (populator *pvcPopulator) Sync() {
+			// Therefore explicitly allow flex volumes to resize
+			volumePlugin, err := populator.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
+			if (err != nil || volumePlugin == nil) && pvcStatusSize.Cmp(pvcSize) < 0 {
+				err = fmt.Errorf("didn't find a plugin capable of expanding the volume; " +
```
same here

aniket-s-kulkarni replied:
@chakri-nelluri the same comment as above - there has been no change to this behavior.

I will update event type as mentioned above.

chakri-nelluri replied:
ACK

chakri-nelluri commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -287,3 +291,38 @@ func (plugin *flexVolumePlugin) getDeviceMountPath(spec *volume.Spec) (string, e
 	return path.Join(mountsDir, volumeName), nil
 }
+
+func (plugin *flexVolumePlugin) ExpandVolumeDevice(spec *volume.Spec, newSize resource.Quantity, oldSize resource.Quantity) (resource.Quantity, error) {
```
*Nit - Move these to a separate file to following flex volume file layout format.

aniket-s-kulkarni replied:
Sure.

chakri-nelluri commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -287,3 +291,38 @@ func (plugin *flexVolumePlugin) getDeviceMountPath(spec *volume.Spec) (string, e
+	call.Append(strconv.FormatInt(newSize.Value(), 10))
+	call.Append(strconv.FormatInt(oldSize.Value(), 10))
+
+	_, err = call.Run()
```
if the call is not supported, use a default function and log it.

chakri-nelluri commented on pkg/volume/flexvolume/plugin.go:
```diff
@@ -287,3 +291,38 @@ func (plugin *flexVolumePlugin) getDeviceMountPath(spec *volume.Spec) (string, e
+	call.Append(devicePath)
+	call.Append(deviceMountPath)
+
+	_, err := call.Run()
```
Same here, use a default function and log it. All other calls do it today.

aniket-s-kulkarni replied:
Sure

chakri-nelluri replied:
Also, for expandFS, we should use the default EXT/XFS expander if the call is not supported. Some plugins wouldn't want to implement expandFS and we can use the default EXT/XFS expander for them.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1265,26 +1281,53 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
+
+	if volumePlugin == nil && !volumeSpec.IsControllerExpandable() {
 		return volumetypes.GeneratedOperations{}, fmt.Errorf("Can not find plugin for expanding volume: %q", pvcWithResizeRequest.QualifiedName())
 	}
```
Can this entire if-else be solved by a instance of flexvolume plugin that is returned via `FindExapndablePlugin` call and performs a no-op resize when called and returns with success response?

aniket-s-kulkarni replied:
@gnufied to install a fake/no-op flexvolume plugin on the controller that will be found via `FindExpandablePlugin` for any driver name in the Spec? If we were to do that we'd have to do something like `if Spec.IsKubeletExpandable() { return noopExpandableInstance{} }` 

If that's fine I will make that change.

aniket-s-kulkarni commented:
@chakri-nelluri @gnufied I have addressed all comments for above with the exception of one - adding a _default_ implementation of `ExpandFS` call for existing intree plugins.

I will now add that and push that in as a separate commit in this PR. Meanwhile, maybe you can take a look at rest of the changes.

aniket-s-kulkarni commented:
@gnufied would it be acceptable if - 

1. any of the intree plugin implements `ExpandFS` method then that method is called (since the plugin becomes a `volume.FSResizableVolumePlugin`)
2. if they don't, then the control flow continues onto generic fs resize like today

I think this will achieve your goal without needing to add noop `ExpandFS` methods to all in-tree plugins right now

aniket-s-kulkarni commented:
> @gnufied would it be acceptable if -
> 
> any of the intree plugin implements ExpandFS method then that method is called (since the plugin becomes a volume.FSResizableVolumePlugin)
> if they don't, then the control flow continues onto generic fs resize like today
> I think this will achieve your goal without needing to add noop ExpandFS methods to all in-tree plugins right now

@chakri-nelluri @gnufied I have added the functionality. You can now review the changes.

aniket-s-kulkarni commented:
/retest

gnufied commented on pkg/volume/plugins.go:
```diff
@@ -909,3 +979,57 @@ type dummyPluginProber struct{}
 func (*dummyPluginProber) Init() error                  { return nil }
 func (*dummyPluginProber) Probe() ([]ProbeEvent, error) { return nil, nil }
+
+type noopExpandableVolumePluginInstance struct {
```
Can we move this new type into its own file?

aniket-s-kulkarni replied:
@gnufied Sure I can do that.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -632,13 +632,36 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, devi
+			var resizeStatus bool
+			var resizeErr error
+
+			fsResizablePlugin, callExpandFS := expandableVolumePlugin.(volume.FSResizableVolumePlugin)
```
Didn't we discuss that `ExpandFS` call should be implemented by all plugins that want to expand file system? In most cases, the call with do nothing other than call out to resize util function but it will enable more polymorphic behaviour.

aniket-s-kulkarni replied:
@gnufied agree and we had two options 

1. Add `ExpandFS` to all in-tree plugins right away and then add boilerplate code in there which calls `resizefs.NewResizer()` etc 
2. Do a lazy approach where if a plugin hasn't implemented that method, `resizefs.NewResizer()...` gets called at the operationGenerator level by default, and if a plugin _does_ implement `ExpandFS` method then it automatically will get called - isn't that polymorphic.

I can go with #1 if you want me to, but don't you think we are being polymorphic as intended?

chakri-nelluri replied:
+1 for @gnufied. Each plugin should have this call.

aniket-s-kulkarni replied:
@gnufied @chakri-nelluri OK that's fine. I will change it and go with #1

gnufied commented on test/e2e/storage/flexvolume_resize.go:
```diff
@@ -0,0 +1,169 @@
+		err = waitForControllerVolumeResize(pvc, c)
+		Expect(err).NotTo(HaveOccurred(), "While waiting for pvc resize to finish")
+
+		By("Waiting for file system resize to finish")
```
I thought fs resize by default will only happen when pod is restarted. How come in this case volume is being resized without waiting for it? The online expansion feature is still in Alpha.

aniket-s-kulkarni replied:
@gnufied this test is for online resize. So perhaps I can add a Skip unless `-feature-gates=ExpandPersistentVolumes=true,ExpandInUsePersistentVolumes=true` is passed to the test.

gnufied replied:
So now that we don't depend on ONLINE resizing, don't we need to restart the pod for flex volume?

aniket-s-kulkarni replied:
> So now that we don't depend on ONLINE resizing, don't we need to restart the pod for flex volume?

1. basically if the online expansion feature gate is ON, then the resizing should be done w/o pod restarts
2. else it should be _after_ a pod restarts.

For #2 tests already exist. For #1 I'm adding the new test in this PR that's why its not expecting the pod to restart.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -632,13 +632,36 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, devi
+				// if any plugin implements FSResizableVolumePlugin interface then call its ExpandFS method
+				// before that if the volume is "kubelet expandable" - like a flex volume, then first call
+				// its ExpandVolumeDevice method before procceeding to ExpandFS
+				if volumeToMount.VolumeSpec.IsKubeletExpandable() {
```
What happens if volume type does not support `ExpandVolumeDevice` when volume is mounted?

aniket-s-kulkarni replied:
@gnufied 

1. If a kubelet expandable volume (i.e. Flex) doesn't support `ExpandVolumeDevice` at all, then as you have already suggested, its dynamic provisioner storage class will set `allowVolumeExpansion: false` and this issue is addressed
2. But if it sets `allowVolumeExpansion: true` and then returns `unsupported` from `ExpandVolumeDevice` then its probably a driver issue that needs to be addressed right

chakri-nelluri replied:
That will make it tricky, today we only care about volumes in use. If a volume is not mounted, we don't even know which node to run this from. That should be fixed with CSI.

aniket-s-kulkarni replied:
@gnufied @chakri-nelluri OK I will go with #1 and amend the commit.

aniket-s-kulkarni replied:
@chakri-nelluri but this code will only be triggered for volumes that are in use. The reconciler checks that the volumes are in use and are mounted (and the ExpandInUsePersistentVolumes feature gate is enabled) before the code flow reaches here.

So am I missing a case here?

chakri-nelluri replied:
What I meant is - "ExpandVolumeDevice when volume is mounted?" cannot be done and the plugin has to fail.

gnufied replied:
If we agree to make flexvolume resize online only, that is fine. But a flexvolume plugin author must know that, it only support online volume resizing.  cc @saad-ali

gnufied replied:
Why is `IsKubeletExpandable` check require in controller side for flex-volume plugins? Can you elaborate?

aniket-s-kulkarni replied:
@gnufied the `IsKubeletExpandable` check is **not** required on the controller side for flex volumes. It **is** required on the kubelet side to ensure that we call `ExpandVolumeDevice` method for flex volumes first (in `resizeFileSystem` method) _before_ calling the `ExpandFS` method.

gnufied reviewed (commented):
Looks better. thank you.

chakri-nelluri commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -217,7 +217,11 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
-				expc.recorder.Event(newPVC, v1.EventTypeNormal, events.ExternalExpanding,
+				eventType := v1.EventTypeNormal
+				if err != nil {
+					eventType = v1.EventTypeWarning
```
Change this to EventTypeError?

chakri-nelluri commented on pkg/controller/volume/expand/pvc_populator.go:
```diff
@@ -99,15 +99,19 @@ func (populator *pvcPopulator) Sync() {
-				populator.recorder.Event(pvc, v1.EventTypeNormal, events.ExternalExpanding,
+				eventType := v1.EventTypeNormal
+				if err != nil {
+					eventType = v1.EventTypeWarning
```
Same here

chakri-nelluri commented on pkg/volume/flexvolume/expander.go:
```diff
@@ -0,0 +1,69 @@
+			size = strconv.FormatInt(newSize.Value(), 10)
+		}
+	}
+	call.Append(size)
```
this can't be zero. Do we support resize on inline volumes? If not, just return error if spec.PersistentVolume != nil.

aniket-s-kulkarni replied:
@chakri-nelluri OK will do.

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -394,28 +394,28 @@ func (spec *Spec) Name() string {
 // IsControllerExpandable returns true for volume types that can be expanded by the controller
 // and false for those that need to be run on the node. Flex volume is the only one in the
 // latter category since its typically not installed on the controller
```
Please fix typo - change 'its' to 'it is'

gnufied commented on pkg/controller/volume/expand/expand_controller.go:
```diff
@@ -208,16 +208,25 @@ func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
-			glog.V(3).Infof("Ignoring the PVC %q (uid: %q) : %v.",
-				util.GetPersistentVolumeClaimQualifiedName(newPVC), newPVC.UID, err)
-			return
+		if volumeSpec := volume.NewSpecFromPersistentVolume(pv, false); volumeSpec.IsKubeletExpandable() {
```
ideally - all these changes should be behind a feature gate.

gnufied commented:
@aniket-s-kulkarni one problem I see is, lot of these changes are spilling over into code flow that is beta and supported. Most code changes for this should be behind feature gate or should be written in such a way that, they don't under beta code flow.

aniket-s-kulkarni commented:
@chakri-nelluri @gnufied I have made changes as mentioned in the review comments. Can you take a look pl?

> @aniket-s-kulkarni one problem I see is, lot of these changes are spilling over into code flow that is beta and supported. Most code changes for this should be behind feature gate or should be written in such a way that, they don't affect regular code flow

@gnufied the control flow for this is indeed behind the feature gates. Neither the faux expansion on the controller nor the kubelet expansion or expandFS would be called for a volume unless `ExpandInUsePersistentVolumes` featuregate is set. It is checked in the controller threads and the kubelet reconciler as well. 

In addition, I can add that feature gate check in `IsKubeletExpandable()` method as well - is that what you would suggest?

aniket-s-kulkarni commented:
@gnufied @chakri-nelluri 

- I have added all the changes requested so far
- The e2e tests have been modified to have the feature gate check
- Also the controller side code (`IsKubeletExpandable`) is also behind the feature gate as an added layer of protection

can you do review?

Secondly, now that this would be targeted for 1.13 what if any change needs to be done to this PR?

aniket-s-kulkarni commented:
@gnufied @chakri-nelluri could you please review this ?

aniket-s-kulkarni commented:
/retest pull-kubernetes-integration

aniket-s-kulkarni commented:
/retest

kacole2 commented:
@aniket-s-kulkarni is this feature targeting 1.13?  1.13 release is targeted to be more ‘stable’ and will have an aggressive timeline. Please only include this enhancement in the milestone if there is a high level of confidence it will meet the following deadlines:

- Docs (open placeholder PRs): 11/8
- Code Slush: 11/9
- Code Freeze Begins: 11/15
- Docs Complete and Reviewed: 11/27

aniket-s-kulkarni commented:
@kacole2 I understand. @chakri-nelluri and @gnufied are going to review it this week.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -632,13 +630,31 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, devi
+					resizeStatus = true
+				}
+			}
+			if !callExpandFS || util.IsGenericResizeRequiredError(resizeErr) {
```
hrm, flexvolume plugin implementation details should not leak into operation_generator.go . I am sure if plugin did not provide an `ExpandFS` implementation, the default implementation can be called from flexvolume plugin code itself rather than branching out here.

aniket-s-kulkarni replied:
@gnufied OK fair enough. I will make sure the defaults implementation of flexvolume as well as others, calls generic resize directly.

aniket-s-kulkarni commented:
/retest

chakri-nelluri commented on pkg/volume/flexvolume/expander.go:
```diff
@@ -0,0 +1,74 @@
+	if !ok {
+		return fmt.Errorf("Capacity not found in PersistentVolume for spec: %s", spec.Name())
+	}
+	call.Append(strconv.FormatInt(newSize.Value(), 10))
```
Adding newSize twice here and above?

aniket-s-kulkarni replied:
Yes, I will fix that.

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -388,6 +395,36 @@ func (spec *Spec) Name() string {
+// IsKubeletExpandable returns true for volume types that can be expanded only by the node
+// and not the controller. Currently Flex volume is the only one in this category since
+// it is typically not installed on the controller
+func (spec *Spec) IsKubeletExpandable() bool {
```
Remove this?

chakri-nelluri commented on pkg/volume/flexvolume/expander.go:
```diff
@@ -0,0 +1,74 @@
+	call.Append(strconv.FormatInt(newSize.Value(), 10))
+	call.Append(strconv.FormatInt(oldSize.Value(), 10))
+
+	if spec.PersistentVolume == nil {
```
Move this check to top.

chakri-nelluri commented on pkg/volume/plugins.go:
```diff
@@ -388,6 +395,36 @@ func (spec *Spec) Name() string {
+// this is used in context on the controller where the plugin lookup fails
+// as volume expansion on controller isn't supported, but a plugin name is
+// required
+func (spec *Spec) KubeletExpandablePluginName() string {
```
This is completely flexvolume specific code in core layer. Anyway to avoid it?

chakri-nelluri commented:
Thanks @aniket-s-kulkarni Added minor comments..

aniket-s-kulkarni commented:
> Thanks @aniket-s-kulkarni Added minor comments..

@chakri-nelluri I have addressed the comments. Do you have any other comments or are you OK with the rest of the PR.

chakri-nelluri commented:
Thanks @aniket-s-kulkarni. Flex changes LGTM.
/LGTM

chakri-nelluri commented:
/LGTM

gnufied commented on test/e2e/storage/flexvolume_resize.go:
```diff
@@ -0,0 +1,178 @@
+	return stKlass, err
+}
+
+var _ = utils.SIGDescribe("Mounted flexvolume volume expand [Slow] [Feature:ExpandInUsePersistentVolumes]", func() {
```
This e2e should not have the feature tag since this feature does not tie itself to ONLINE resizing.

Commit:
Incorporating review comments

chakri-nelluri commented:
/LGTM

Commit:
Updating build files after review comments

gnufied commented:
/approve

gnufied commented:
/lgtm

aniket-s-kulkarni commented:
/assign @saad-ali 

@saad-ali assigning you as approver, since this has a test case change that needs approval.

chakri-nelluri commented:
/LGTM

gnufied commented:
Please don't forget to open follow up PR that adds e2e that does not depend on online resizing feature.

gnufied commented on pkg/volume/flexvolume/expander.go:
```diff
@@ -0,0 +1,67 @@
+	"strconv"
+)
+
+func (plugin *flexVolumePlugin) ExpandVolumeDevice(spec *volume.Spec, newSize resource.Quantity, oldSize resource.Quantity) (resource.Quantity, error) {
```
Will this implementation EVER get called? It looks like, even if flex volume plugin is installed on the controller the code still returns a no-op version of flex plugin.

gnufied replied:
nvm. looks like it works as expected. thanks.

gnufied commented:
Putting the PR on hold for a bit.
/hold

gnufied commented:
/hold cancel

saad-ali reviewed (commented):
/approve
